### PR TITLE
Port changes to master from release branch 6_4

### PR DESCRIPTION
**Pull request by lukedegruchy**
- Restore changes from this merge request. that were lost in master:  https://github.com/hapifhir/hapi-fhir/pull/4591

Closes https://github.com/hapifhir/hapi-fhir/issues/4786

### CODE REVIEW

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/bulk/export/ExpandResourcesStepJpaTest.java [hunk 1]
```diff
@@ -85,10 +85,10 @@ public void testBulkExportExpandResourcesStep(JpaStorageSettings.TagStorageModeE
...
-		assertThat(expandedResourceList.getStringifiedResources().get(0), containsString("{\"system\":\"http://dynamic\",\"code\":\"tag0\"}"));
-		assertThat(expandedResourceList.getStringifiedResources().get(1), containsString("{\"system\":\"http://static\",\"code\":\"tag\"}"));
-		assertThat(expandedResourceList.getStringifiedResources().get(1), containsString("{\"system\":\"http://dynamic\",\"code\":\"tag1\"}"));
+		assertThat(expandedResourceList.getStringifiedResources().get(0), containsString("{\"system\":\"http://static\",\"version\":\"1\",\"code\":\"tag\",\"userSelected\":true}"));
+		assertThat(expandedResourceList.getStringifiedResources().get(0), containsString("{\"system\":\"http://static\",\"version\":\"1\",\"code\":\"tag\",\"userSelected\":true}"));
+		assertThat(expandedResourceList.getStringifiedResources().get(1), containsString("{\"system\":\"http://static\",\"version\":\"1\",\"code\":\"tag\",\"userSelected\":true}"));
+		assertThat(expandedResourceList.getStringifiedResources().get(1), containsString("{\"system\":\"http://static\",\"version\":\"1\",\"code\":\"tag\",\"userSelected\":true}"));
```
Comment (Qingyixia) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/bulk/export/ExpandResourcesStepJpaTest.java:91]: Question: Just wondering why we need identical assert statements here.

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4MetaTest.java [hunk 2]
```diff
@@ -142,6 +153,163 @@ public void testAddTagAndSecurityLabelWithSameValues() {
...
+					.filter(resourceHistoryTag -> expectedVersion1.equals(resourceHistoryTag.getTag().getVersion()))
+					.filter(resourceHistoryTag -> expectedUserSelected1 == resourceHistoryTag.getTag().getUserSelected())
+					.count()),
+				() -> assertEquals(theExpectedHistoryCounts.theTagV2Qty, theHistoryTags.stream()
+					.filter(resourceHistoryTag -> expectedSystem2.equals(resourceHistoryTag.getTag().getSystem()))
+					.filter(resourceHistoryTag -> expectedCode2.equals(resourceHistoryTag.getTag().getCode()))
+					.filter(resourceHistoryTag -> expectedDisplay2.equals(resourceHistoryTag.getTag().getDisplay()))
+					.filter(resourceHistoryTag -> expectedVersion2.equals(resourceHistoryTag.getTag().getVersion()))
+					.filter(resourceHistoryTag -> expectedUserSelected2 == resourceHistoryTag.getTag().getUserSelected())
+					.count())
```
Comment (Qingyixia) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4MetaTest.java:306]: Great test! They are so organized.

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/bulk/export/ExpandResourcesStepJpaTest.java (hunk 1 above)
Comment (lukedegruchy) [reply hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/bulk/export/ExpandResourcesStepJpaTest.java:91]: Good point.  I'll remove the dupes.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/migrate/tasks/HapiFhirJpaMigrationTasks.java [hunk 3]
```diff
@@ -324,6 +325,38 @@ protected void init660() {
...
+			addTagDefConstraint.put(DriverTypeEnum.MARIADB_10_1, "ALTER TABLE HFJ_TAG_DEF ADD CONSTRAINT IDX_TAGDEF_TYPESYSCODEVERUS UNIQUE (TAG_TYPE, TAG_CODE, TAG_SYSTEM, TAG_VERSION, TAG_USER_SELECTED)");
+			addTagDefConstraint.put(DriverTypeEnum.MSSQL_2012, "ALTER TABLE HFJ_TAG_DEF ADD CONSTRAINT IDX_TAGDEF_TYPESYSCODEVERUS UNIQUE (TAG_TYPE, TAG_CODE, TAG_SYSTEM, TAG_VERSION, TAG_USER_SELECTED)");
+			addTagDefConstraint.put(DriverTypeEnum.MYSQL_5_7, "ALTER TABLE HFJ_TAG_DEF ADD CONSTRAINT IDX_TAGDEF_TYPESYSCODEVERUS UNIQUE (TAG_TYPE, TAG_CODE, TAG_SYSTEM, TAG_VERSION, TAG_USER_SELECTED)");
+			addTagDefConstraint.put(DriverTypeEnum.ORACLE_12C, "ALTER TABLE HFJ_TAG_DEF ADD CONSTRAINT IDX_TAGDEF_TYPESYSCODEVERUS UNIQUE (TAG_TYPE, TAG_CODE, TAG_SYSTEM, TAG_VERSION, TAG_USER_SELECTED)");
+			addTagDefConstraint.put(DriverTypeEnum.POSTGRES_9_4, "ALTER TABLE HFJ_TAG_DEF ADD CONSTRAINT IDX_TAGDEF_TYPESYSCODEVERUS UNIQUE (TAG_TYPE, TAG_CODE, TAG_SYSTEM, TAG_VERSION, TAG_USER_SELECTED)");
+			version.executeRawSql("20230209.5", addTagDefConstraint);
+		}
+	}
+
+	protected void init630() {
```
Comment (tadgh) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/migrate/tasks/HapiFhirJpaMigrationTasks.java:359]: issue: I think this is an error. Please remove the init630() chunk, and move this into the init640() chunk that aleady existed.

File: hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/entity/TagDefinition.java [hunk 4]
```diff
@@ -132,6 +147,31 @@ public void setTagType(TagTypeEnum theTagType) {
...
+
+	private void setVersionAfterTrim(String theVersion) {
+		if (theVersion != null) {
+			myVersion = StringUtils.truncate(theVersion, 30);
+		}
+	}
+
+	public Boolean getUserSelected() {
+		// TODO: LD: this is not ideal as we are implicitly assuming null is false.
+		//  Ideally we should fix IBaseCoding to return wrapper Boolean but that will involve another core/hapi release
```
Comment (tadgh) [inline hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/entity/TagDefinition.java:166]: question: I imagine this todo was ported and is not new. Just checking

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/bulk/export/ExpandResourcesStepJpaTest.java (hunk 1 above)
Comment (tadgh) [reply hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/bulk/export/ExpandResourcesStepJpaTest.java:89]: oddly those are not dupes. The first checks element 0, and the second checks element 1 of the stringified resources. That said... probably overkill anyhow.

**Review by tadgh — APPROVED** [review]
Approved pending minor things. Recommend compiling this against CDR before merging.

File: hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/entity/TagDefinition.java (hunk 4 above)
Comment (lukedegruchy) [reply hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/entity/TagDefinition.java:166]: Yes, I was the one who wrote the initial TODO after you and I decided the add this hack at the last minute to the 2nd Feb release.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/migrate/tasks/HapiFhirJpaMigrationTasks.java (hunk 3 above)
Comment (lukedegruchy) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/migrate/tasks/HapiFhirJpaMigrationTasks.java:359]: Are you sure we should be running this migration on 6_3_0?   This is what I see on master:

```
	protected void init640() {
		Builder version = forVersion(VersionEnum.V6_3_0);
```

**Pull request merged**